crates/oxc_transformer/src/es2018/object_spread.rs

```diff
@@ -0,0 +1,139 @@
+//! * Object rest/spread TC39 proposal: <https://github.com/tc39/proposal-object-rest-spread>
+
+use super::object_rest_spread::ObjectRestSpreadOptions;
+use crate::context::Ctx;
+
+use oxc_ast::ast::*;
+use oxc_semantic::ReferenceFlags;
+use oxc_span::SPAN;
+use oxc_traverse::{Traverse, TraverseCtx};
+
+pub struct ObjectSpread<'a> {
+    _ctx: Ctx<'a>,
+    options: ObjectRestSpreadOptions,
+}
+
+impl<'a> ObjectSpread<'a> {
+    pub fn new(options: ObjectRestSpreadOptions, ctx: Ctx<'a>) -> Self {
+        Self { _ctx: ctx, options }
+    }
+}
+impl<'a> Traverse<'a> for ObjectSpread<'a> {
+    fn enter_expression(&mut self, expr: &mut Expression<'a>, ctx: &mut TraverseCtx<'a>) {
+        let Expression::ObjectExpression(obj_expr) = expr else {
+            return;
+        };
+
+        if obj_expr
+            .properties
+            .iter()
+            .all(|prop| matches!(prop, ObjectPropertyKind::ObjectProperty(..)))
+        {
+            return;
+        }
+
+        // collect `y` and `z` from `{ ...x, y, z }`
+        let mut obj_prop_list = ctx.ast.vec();
+        while obj_expr
+            .properties
+            .last()
+            .map_or(false, |prop| matches!(prop, ObjectPropertyKind::ObjectProperty(..)))
+        {
+            let prop = obj_expr.properties.pop().unwrap();
+            obj_prop_list.push(prop);
+        }
+
+        let Some(ObjectPropertyKind::SpreadProperty(mut spread_prop)) = obj_expr.properties.pop()
+        else {
+            unreachable!();
+        };
+
+        self.enter_expression(expr, ctx);
+        let mut arguments = ctx.ast.vec();
+        arguments.push(Argument::from(ctx.ast.move_expression(expr)));
+        arguments.push(Argument::from(ctx.ast.move_expression(&mut spread_prop.argument)));
+
+        let callee = self.get_extend_object_callee(ctx);
+
+        // ({ ...x }) => _objectSpread({}, x)
+        *expr = ctx.ast.expression_call(
+            SPAN,
+            callee,
+            None::<TSTypeParameterInstantiation>,
+            arguments,
+            false,
+        );
+
+        // ({ ...x, y, z }) => _objectSpread(_objectSpread({}, x), { y, z });
+        if !obj_prop_list.is_empty() {
+            obj_prop_list.reverse();
+            let mut arguments = ctx.ast.vec();
+            arguments.push(Argument::from(ctx.ast.move_expression(expr)));
+            arguments.push(Argument::from(ctx.ast.expression_object(SPAN, obj_prop_list, None)));
+
+            let callee = self.get_extend_object_callee(ctx);
+
+            *expr = ctx.ast.expression_call(
+                SPAN,
+                callee,
+                None::<TSTypeParameterInstantiation>,
+                arguments,
+                false,
+            );
+        }
+    }
+}
+
+impl<'a> ObjectSpread<'a> {
+    fn object_assign(ctx: &mut TraverseCtx<'a>) -> Expression<'a> {
+        let ident =
+            ctx.create_reference_id(SPAN, ctx.ast.atom("Object"), None, ReferenceFlags::Read);
```

Comment:
If I've understood correctly, I should avoid using a user-defined `Object` and instead refer to the global `Object`, correct?
If the user has created their own `Object` variable, I should rename it, perhaps to something like `Object2`.

Reply:
I think we can just accept that if the source code has a variable called `Object`, the transformed code is not going to work properly. I think that's OK - overriding globals is asking for trouble, so don't complain if you get it!

My point is not that we need to fix that, but that our scope tree should completely match the AST. So if we insert `Object.assign` and `Object` in that position resolves to a local variable, the scope tree should correctly reflect that.

PS: If we *really* wanted to tackle this, we could use `{}.constructor.assign` instead of `Object.assign` where there's an existing binding for `Object`. But even that's not infallible - what if the user overwrote `Object.prototype.constructor`? So I don't think it's worthwhile trying to handle. Ultimately you can't 100% prevent the user from sabotaging you.

Reply:
OK, that makes sense.